commands/command_filter_process.go

```diff
@@ -40,33 +44,83 @@ func filterCommand(cmd *cobra.Command, args []string) {
 		ExitWithError(err)
 	}
 
-	_, err := s.NegotiateCapabilities()
+	caps, err := s.NegotiateCapabilities()
 	if err != nil {
 		ExitWithError(err)
 	}
 
+	var supportsDelay bool
+	for _, cap := range caps {
+		if cap == "capability=delay" {
+			supportsDelay = true
+			break
+		}
+	}
+
 	skip := filterSmudgeSkip || cfg.Os.Bool("GIT_LFS_SKIP_SMUDGE", false)
 	filter := filepathfilter.New(cfg.FetchIncludePaths(), cfg.FetchExcludePaths())
 
+	ptrs := make(map[string]*lfs.Pointer)
+
+	var q *tq.TransferQueue
+	closeOnce := new(sync.Once)
+	available := make(chan *tq.Transfer)
+
+	if supportsDelay {
+		q = tq.NewTransferQueue(tq.Download, getTransferManifest(), cfg.CurrentRemote)
+		go infiniteTransferBuffer(q, available)
+	}
+
 	var malformed []string
 	var malformedOnWindows []string
-
 	for s.Scan() {
 		var n int64
 		var err error
+		var delayed bool
 		var w *git.PktlineWriter
 
 		req := s.Request()
 
-		s.WriteStatus(statusFromErr(nil))
+		if !(req.Header["command"] == "smudge" && req.Header["can-delay"] == "1") && !(req.Header["command"] == "list_available_blobs") {
+			s.WriteStatus(statusFromErr(nil))
+		}
 
 		switch req.Header["command"] {
 		case "clean":
 			w = git.NewPktlineWriter(os.Stdout, cleanFilterBufferCapacity)
 			err = clean(w, req.Payload, req.Header["pathname"], -1)
 		case "smudge":
 			w = git.NewPktlineWriter(os.Stdout, smudgeFilterBufferCapacity)
-			n, err = smudge(w, req.Payload, req.Header["pathname"], skip, filter)
+			if req.Header["can-delay"] == "1" {
+				var ptr *lfs.Pointer
+
+				n, delayed, ptr, err = delayedSmudge(s, w, req.Payload, q, req.Header["pathname"], skip, filter)
+
+				if delayed {
+					ptrs[req.Header["pathname"]] = ptr
+				}
+			} else {
+				from, err := incomingOrCached(req.Payload, ptrs[req.Header["pathname"]])
+				if err != nil {
+					break
+				}
+
+				n, err = smudge(w, from, req.Header["pathname"], skip, filter)
+
+				if err == nil {
+					delete(ptrs, req.Header["pathname"])
+				}
+			}
+		case "list_available_blobs":
+			closeOnce.Do(func() { go q.Wait() })
```

Comment:
🤔 not quite -- I think the confusion has to do with the `readAvailable()` func. `readAvailable` does as many reads as possible before `<-tq.Watch()` becomes blocking, responding to a `list_available_blobs` command with any blobs that it currently has ready, or blocking until at least one blob is ready.

The process looks like this:

1. Git sends all pointers through `command=smudge` with `can-delay=1`.
2. Git issues a `list_available_blobs` command, indicating that no more _new_ blobs will be sent as above. The transfer queue is accordingly closed.
3. We read as many currently available blobs as we can before blocking, or block until a) at least one blob becomes available, or b) the transfer queue has no more blobs to send, in which case we send an empty list, and the process is complete.
4. Git send a `command=smudge` with `can-delay=0` for each blob in the list above.
5. Git issues another `list_available_blobs` command.
6. See step 3.

I did my best to clarify this in 8816463.